cv32e40s/tests/programs/custom/pma/pma.c

```diff
@@ -85,345 +104,220 @@ static void assert_or_die(uint32_t actual, uint32_t expect, char *msg) {
+  printf("pma: Load-reserved to disallowed regions raises precise exception\n");
   reset_volatiles();
-  /* TODO enable when RTL is implemented
   __asm__ volatile("lr.w %0, 0(%1)" : "=r"(tmp) : "r"(IO_ADDR));
-  assert_or_die(mcause, EXCEPTION_LOAD_ACCESS_FAULT, "error: load-reserved to non-atomic should except\n");
-  assert_or_die(mepc, IO_ADDR, "error: load-reserved to non-atomic unexpected mepc\n");
-  assert_or_die(mtval, MTVAL_READ, "error: load-reserved to non-atomic unexpected mtval\n");
-  */
+  assert_or_die(g_mcause, EXCEPTION_LOAD_ACCESS_FAULT, "error: load-reserved to non-atomic should except\n");
+  assert_or_die(g_mepc, IO_ADDR, "error: load-reserved to non-atomic unexpected mepc\n");
+  assert_or_die(g_mtval, MTVAL_READ, "error: load-reserved to non-atomic unexpected mtval\n");
 
-  // Store-conditional to disallowed regions raises precise exception
+  printf("pma: Store-conditional to disallowed regions raises precise exception\n");
   reset_volatiles();
-  /* TODO enable when RTL is implemented
   __asm__ volatile("sc.w %0, %0, 0(%1)" : "=r"(tmp) : "r"(IO_ADDR));
-  assert_or_die(mcause, EXCEPTION_STOREAMO_ACCESS_FAULT, "error: store-conditional to non-atomic should except\n");
-  assert_or_die(mepc, IO_ADDR, "error: store-conditional to non-atomic unexpected mepc\n");
-  assert_or_die(mtval, MTVAL_READ, "error: store-conditional to non-atomic unexpected mtval\n");
-  */
+  assert_or_die(g_mcause, EXCEPTION_STOREAMO_ACCESS_FAULT, "error: store-conditional to non-atomic should except\n");
+  assert_or_die(g_mepc, IO_ADDR, "error: store-conditional to non-atomic unexpected mepc\n");
+  assert_or_die(g_mtval, MTVAL_READ, "error: store-conditional to non-atomic unexpected mtval\n");
+*/
 
 
-  // Check Zce-related PMA features
-
-  // Push instrs should fault to IO but pass for MEM
-  check_zce_push();
-
-  // Pop instrs should fault to IO but pass for MEM
-  check_zce_pop();
-
-  // Table jump failing first fetch should be the fault of the table jump
-  reset_volatiles();
-  tmp = fail_first_tblj();
-  /* TODO enable when RTL is implemented
-  assert_or_die(mcause, EXCEPTION_INSN_ACCESS_FAULT, "error: tblj failing first should instruction access fault\n");
-  assert_or_die(mepc, tmp, "error: tblj first expected different mepc\n");
-  assert_or_die(mtval, MTBLJALVEC, "error: tblj first expected different mtval\n");
-  */
-
-  // Table jump failing second fetch should be the fault of the target
-  reset_volatiles();
-  /* TODO enable when RTL is implemented
-  make sure table is executable but target is not
-  __asm__ volatile("c.tbljal 1");
-  assert_or_die(mcause, EXCEPTION_INSN_ACCESS_FAULT, "error: tblj failing second should instruction access fault\n");
-  assert_or_die(mepc, TBLJ_TARGET_ADDR, "error: tblj second expected different mepc\n");
-  assert_or_die(mtval, MTVAL_READ, "error: tblj second expected different mtval\n");
-  */
+  // Zc features are in the vplan, but are much easier to check with assertions.
```

Comment:
I decided to not do the pma-zc checking in C/asm because it gets complicated by the "observer effect" (_"the act of observing will influence the phenomenon being observed"_).
With an assert you can get the given scenario and observe that the outcome is as expected, without the additional constraint that the system state maintains a healthy SW control flow.